### PR TITLE
Replace list of outstanding callback requests with count

### PR DESCRIPTION
**Pull request by vjpai**
We kept a list of callback requests for proper  handling at shutdown but that's vestigial at this point; instead a count is sufficient and avoids a lot of locks and contention.

Performance goes from 820K (after #17901) to 1.2M QPS on 62-channel 0-byte test

### CODE REVIEW

File: include/grpcpp/server.h [hunk 1]
```diff
@@ -248,22 +249,15 @@ class Server : public ServerInterface, private GrpcLibraryCodegen {
...
-  };
-  std::vector<MethodReqList*> callback_reqs_;
+  // Outstanding unmatched callback requests, indexed by method.
+  // NOTE: Using a gpr_atm rather than atomic_int because atomic_int isn't
+  //       copyable or movable and thus will cause compilation errors. We
+  //       actually only want to extend the vector before the threaded use
+  //       starts, but this is still a limitation.
+  std::vector<gpr_atm> callback_unmatched_reqs_count_;
+
+  // List of callback requests to start when server actually starts
```
Comment (soheilhy) [inline include/grpcpp/server.h]: nit: please end with "."

File: src/cpp/server/server_cc.cc [hunk 2]
```diff
@@ -428,46 +428,31 @@ class Server::CallbackRequest final : public internal::CompletionQueueTag {
...
-                 SOFT_MINIMUM_SPARE_CALLBACK_REQS_PER_METHOD &&
-             req_->server_->callback_reqs_outstanding_ <
-                 SOFT_MAXIMUM_CALLBACK_REQS_OUTSTANDING)) {
-          spawn_new = true;
-        }
+      auto count =
+          static_cast<int>(gpr_atm_no_barrier_fetch_add(
+              &req_->server_
+                   ->callback_unmatched_reqs_count_[req_->method_index_],
+              static_cast<gpr_atm>(-1))) -
```
Comment (soheilhy) [inline src/cpp/server/server_cc.cc]: nit: I believe the cast to `gpr_atm` is not necessary, here and below. no strong preference though.

File: src/cpp/server/server_cc.cc [hunk 3]
```diff
@@ -791,12 +774,11 @@ Server::~Server() {
...
-    // The entries of the method_list should have already been emptied
-    // during Shutdown as each request is failed by Shutdown. Check that
-    // this actually happened.
-    GPR_ASSERT(method_list->reqs_list.empty());
-    delete method_list;
+  for (auto per_method_count : callback_unmatched_reqs_count_) {
+    // There should be no more unmatched callbacks for any method
+    // as each request is failed by Shutdown. Check that this actually
+    // happened
+    GPR_ASSERT(static_cast<int>(per_method_count) == 0);
```
Comment (soheilhy) [inline src/cpp/server/server_cc.cc]: nit: I'd suggest using `gpr_no_barrier_load` here.

File: include/grpcpp/server.h (hunk 1 above)
Comment (vjpai) [reply include/grpcpp/server.h]: Will do.

File: src/cpp/server/server_cc.cc (hunk 2 above)
Comment (vjpai) [reply src/cpp/server/server_cc.cc]: So, I like casting for my own understanding, but I know that in this case the extra words lead to ugly formatting, so I'll skip this.

File: src/cpp/server/server_cc.cc (hunk 3 above)
Comment (vjpai) [reply src/cpp/server/server_cc.cc]: Will do.

**Comment by vjpai** [comment]
Thanks @soheilhy , I think I've addressed all your comments.

**Comment by vjpai** [comment]
Failures: #17335

File: src/cpp/server/server_cc.cc [hunk 4]
```diff
@@ -428,46 +428,31 @@ class Server::CallbackRequest final : public internal::CompletionQueueTag {
...
+        delete req_;
+        return;
       }
-      if (spawn_new) {
-        auto* new_req = new CallbackRequest(req_->server_, req_->req_list_,
+
+      // If this was the last request in the list or it is below the soft
+      // minimum and there are spare requests available, set up a new one.
+      if (count == 0 || (count < SOFT_MINIMUM_SPARE_CALLBACK_REQS_PER_METHOD &&
+                         count < SOFT_MAXIMUM_CALLBACK_REQS_OUTSTANDING)) {
```
Comment (vjpai) [inline src/cpp/server/server_cc.cc]: This was a mistaken change. I'll fix.

**Pull request merged**